LTC_ECCSIG_RFC5656 SSH+ECDSA signature format

rmw42 opened the pull request:
Implementation of SSH+ECDSA signature format for ECC signatures as defined in RFC5656.

Includes a SSH data model encode/decode (gated on #define LTC_SSH), and OID string<->Curve comparison helper function (also used for secp256k1 checks for LTC_ECCSIG_ETH27)

### Checklist

* [x] documentation is added or updated
* [x] tests are added or updated

rmw42 committed:
Initial commit of SSH+ECDSA signature format

Wrap signature format in #ifdef LTC_SSH
Update docs
Code review fixes
Replace strcmp/memcmp with XSTRCMP/XMEMCMP for check-source
Fix for check-defines
XSTRCMP/XMEMCMP != 0
GCC7.3 wants only literal strings for sprintf format
Code review changes
Rework SSH decoding and tests
Fix encoding and tests
COMPARE_TESTVECTOR macro
Single return point in ssh_decode_sequence_multi
Actually use XSTRNCPY rather than just defining it
More code review fixes
Code review tweaks
Ensure it's not possible to read past buffer end
Keep track of size remaining, not end pointer

rmw42 commented:
@karel-m As promised :)

karel-m commented on src/headers/tomcrypt_pk.h:
```diff
@@ -321,6 +324,10 @@ int  ecc_recover_key(const unsigned char *sig,  unsigned long siglen,
                      int recid, ecc_signature_type sigformat, ecc_key *key);
 
+#ifdef LTC_SSH
+int ecc_ssh_ecdsa_encode_name(char *buffer, unsigned long buflen, const ecc_key *key);
```
Please move this part to `tomcrypt_private.h`

rmw42 replied:
OK

karel-m commented on src/misc/ssh/ssh_test.c:
```diff
@@ -0,0 +1,243 @@
+/* LibTomCrypt, modular cryptographic library -- Tom St Denis
```
Please move `src/misc/ssh/ssh_test.c` to `tests/ssh_test.c` (+ do not forget to run `./updatemakes.sh`)

rmw42 replied:
OK

karel-m commented on src/misc/ssh/ssh_test.c:
```diff
@@ -0,0 +1,243 @@
+*/
+int ssh_test(void)
+{
+#ifndef LTC_TEST
```
Once we move this stuff to `tests/ssh_test.c` there is no need for `#ifndef LTC_TEST`

karel-m commented on src/pk/ecc/ecc_cmp_oid_str.c:
```diff
@@ -0,0 +1,38 @@
+
+int ecc_cmp_oid_str(const char *oidstr, const ecc_key *key)
+{
+   char oid[32];
```
Isn't it too small? I am for `oid[64]`.

rmw42 replied:
If an OID can be 64 chars then the SSH 64-char limit on the "ecdsa-sha2-XXXX" string would need to be higher. Happy to change it if there are super-long OIDs out there.

rmw42 replied:
Changed to 64

karel-m commented on src/pk/ecc/ecc_cmp_oid_str.c:
```diff
@@ -0,0 +1,38 @@
+
+   err = pk_oid_num_to_str(key->dp.oid, key->dp.oidlen, oid, &oidlen);
+   if (err != CRYPT_OK) return err;
+   else if (!XSTRCMP(oidstr, oid)) return CRYPT_OK;
```
I prefer to use explicit: `XSTRCMP(x, y) != 0` or `XSTRCMP(x, y) == 0` (applies also to other places)

karel-m replied:
and clang-tidy complains here:
```
src/pk/ecc/ecc_cmp_oid_str.c:30:4: error: do not use 'else' after 'return' [readability-else-after-return,-warnings-as-errors]
   else if (!XSTRCMP(oidstr, oid)) return CRYPT_OK;
   ^~~~~~~~
```

So maybe change it to: `if (XSTRCMP(oidstr, oid) == 0) return CRYPT_OK;`

rmw42 replied:
Will do. And I'll go with the "goto error / single return" pattern so clang-tidy is happy.

karel-m commented on src/pk/ecc/ecc_ssh_ecdsa_encode_name.c:
```diff
@@ -0,0 +1,57 @@
+   if (ecc_cmp_oid_str("1.2.840.10045.3.1.7", key) == CRYPT_OK) {
+      /* nistp256 - secp256r1 - OID 1.2.840.10045.3.1.7 */
+
+      if (buflen <= 19) { err = CRYPT_BUFFER_OVERFLOW; goto error; }
```
Maybe rather `buflen <= 20` - according to `man 3 strcpy`: The `strcpy()` function copies the string pointed to by src, **including the terminating null byte**.

rmw42 replied:
Yeah, the string is 19 chars long, so we need 19+1 to store it... if the buffer is 19 or fewer, it's too small. I figured "<= 19" was slightly clearer than "< 20" or "< 19 + 1", but none are particularly great.

What about:
```
   const char *name = "ecdsa-sha2-nistp256";
   if (buflen <= strlen(name)) { err = ... }
   XSTRCPY(buffer, name);
```

karel-m commented on src/pk/ecc/ecc_ssh_ecdsa_encode_name.c:
```diff
@@ -0,0 +1,57 @@
+   else if (ecc_cmp_oid_str("1.3.132.0.34", key) == CRYPT_OK) {
+      /* nistp384 - secp384r1 - OID 1.3.132.0.34 */
+
+      if (buflen <= 19) { err = CRYPT_BUFFER_OVERFLOW; goto error; }
```
`buflen <= 20`

karel-m commented on src/pk/ecc/ecc_ssh_ecdsa_encode_name.c:
```diff
@@ -0,0 +1,57 @@
+   else if (ecc_cmp_oid_str("1.3.132.0.35", key) == CRYPT_OK) {
+      /* nistp521 - secp521r1 - OID 1.3.132.0.35 */
+
+      if (buflen <= 19) { err = CRYPT_BUFFER_OVERFLOW; goto error; }
```
`buflen <= 20`

karel-m commented on src/pk/ecc/ecc_ssh_ecdsa_encode_name.c:
```diff
@@ -0,0 +1,57 @@
+      /* nistp521 - secp521r1 - OID 1.3.132.0.35 */
+
+      if (buflen <= 19) { err = CRYPT_BUFFER_OVERFLOW; goto error; }
+      strcpy(buffer, "ecdsa-sha2-nistp512");
```
I thing it should be `nistp521`

rmw42 replied:
Yeah... my fingers are too familiar with powers of 2 ^_^;;

karel-m commented on src/pk/ecc/ecc_ssh_ecdsa_encode_name.c:
```diff
@@ -0,0 +1,57 @@
+      unsigned long oidlen = 32;
+      if ((err = ecc_get_oid_str(oidstr, &oidlen, key)) != CRYPT_OK) goto error;
+
+      if (buflen <= strlen(oidstr)+11) { err = CRYPT_BUFFER_OVERFLOW; goto error; }
```
Perhaps `if (buflen <= strlen(oidstr)+12)` or `if (buflen <= oidlen + 12)` - sprintf AFAIK writes terminating nul byte

rmw42 replied:
Completely rewritten to use snprintf() and not need any hand-counted string lengths

karel-m commented on src/pk/ecc/ecc_ssh_ecdsa_encode_name.c:
```diff
@@ -0,0 +1,57 @@
+      strcpy(buffer, "ecdsa-sha2-nistp512");
+   }
+   else { /* Otherwise, we use the OID of the curve */
+      char oidstr[32];
```
Isn't it too small? I am for oidstr[64].

rmw42 replied:
Changed to 64

karel-m commented on src/pk/ecc/ecc_ssh_ecdsa_encode_name.c:
```diff
@@ -0,0 +1,57 @@
+   Russ Williams
+*/
+
+int ecc_ssh_ecdsa_encode_name(char *buffer, unsigned long buflen, const ecc_key *key)
```
To keep the style used on other places in libtomcrypt it should be `unsigned long *buflen` (but as it is a private function maybe we can leave it as it is, @sjaeckel ?)

rmw42 replied:
Done, and it writes out the length into *buflen.

karel-m commented:
Great job!

karel-m commented:
And 2 warnings:
```
src/misc/ssh/ssh_test.c:59:5: error: no previous prototype for ‘_ssh_encoding_test’ [-Werror=missing-prototypes]
 int _ssh_encoding_test(void)
     ^~~~~~~~~~~~~~~~~~
src/misc/ssh/ssh_test.c:154:5: error: no previous prototype for ‘_ssh_decoding_test’ [-Werror=missing-prototypes]
 int _ssh_decoding_test(void)
     ^~~~~~~~~~~~~~~~~~
```

karel-m commented on src/misc/ssh/ssh_test.c:
```diff
@@ -0,0 +1,243 @@
+  LTC_SSH encoding test
+  @return CRYPT_OK if successful
+*/
+int _ssh_encoding_test(void)
```
`static int _ssh_encoding_test(void)`

rmw42 replied:
Done.

karel-m commented on src/misc/ssh/ssh_test.c:
```diff
@@ -0,0 +1,243 @@
+  LTC_SSH decoding test
+  @return CRYPT_OK if successful
+*/
+int _ssh_decoding_test(void)
```
`static int _ssh_decoding_test(void)`

karel-m commented on src/misc/ssh/ssh_test.c:
```diff
@@ -0,0 +1,243 @@
+   DO(ssh_encode_sequence_multi(buffer, &buflen,
+                                LTC_SSHDATA_UINT32, 0x29b7f4aa,
+                                LTC_SSHDATA_EOL,    NULL));
+   if (XMEMCMP(buffer, uint32, buflen)) return CRYPT_FAIL_TESTVECTOR;
```
`if (XMEMCMP(buffer, uint32, buflen) != 0)` + all similar places below

rmw42 replied:
OK

karel-m commented on src/pk/ecc/ecc_cmp_oid_str.c:
```diff
@@ -0,0 +1,38 @@
+   err = pk_oid_num_to_str(key->dp.oid, key->dp.oidlen, oid, &oidlen);
+   if (err != CRYPT_OK) return err;
+   else if (!XSTRCMP(oidstr, oid)) return CRYPT_OK;
+   else return CRYPT_INVALID_ARG;
```
without `else`

karel-m commented on src/pk/ecc/ecc_ssh_ecdsa_encode_name.c:
```diff
@@ -0,0 +1,57 @@
+      /* nistp256 - secp256r1 - OID 1.2.840.10045.3.1.7 */
+
+      if (buflen <= 19) { err = CRYPT_BUFFER_OVERFLOW; goto error; }
+      strcpy(buffer, "ecdsa-sha2-nistp256");
```
clang-tidy complains here (+ few other places in this file):
```
src/pk/ecc/ecc_ssh_ecdsa_encode_name.c:32:7: error: Call to function 'strcpy' is insecure as it does not provide bounding of the memory buffer. Replace unbounded copy functions with analogous functions that support length arguments such as 'strlcpy'. CWE-119 [clang-analyzer-security.insecureAPI.strcpy,-warnings-as-errors]
      strcpy(buffer, "ecdsa-sha2-nistp384");
      ^
```
So perhaps use something else than `strcpy` e.g. `strncpy` / `XMEMCPY`

rmw42 replied:
Replaced with snprintf() and combined with buffer length checking, so no more magic numbers.

rmw42 commented:
> Great job!

Cheers! :)

rmw42 commented:
Fixes in, and building now... hopefully that's all acceptable. Let me know if I missed anything! :)

karel-m commented:
gcc-7.3 with `-Wformat=2`
```
src/pk/ecc/ecc_ssh_ecdsa_encode_name.c: In function ‘ecc_ssh_ecdsa_encode_name’:
src/pk/ecc/ecc_ssh_ecdsa_encode_name.c:47:4: error: format not a string literal, argument types not checked [-Werror=format-nonliteral]
    size = snprintf(buffer, *buflen, pattern, oidstr);
    ^~~~
```
Can we somehow avoid this warning?

karel-m commented:
clang-tidy is without warnings 👍

rmw42 commented:
> gcc-7.3 with `-Wformat=2`
> 
> ```
> src/pk/ecc/ecc_ssh_ecdsa_encode_name.c: In function ‘ecc_ssh_ecdsa_encode_name’:
> src/pk/ecc/ecc_ssh_ecdsa_encode_name.c:47:4: error: format not a string literal, argument types not checked [-Werror=format-nonliteral]
>     size = snprintf(buffer, *buflen, pattern, oidstr);
>     ^~~~
> ```
> Can we somehow avoid this warning?

Ah, damn, an I thought I was being so clever changing the format string... well, if it has to be a literal, I can get rid of `pattern` entirely, use four `snprintf()` calls in the different branches of the `if` but leave the buffer length check as-is at the end...

Give me 5 minutes ;-)

sjaeckel commented on doc/crypt.tex:
```diff
@@ -5254,6 +5254,17 @@ \subsection{Key Helper Functions}
+
+\index{ecc\_cmp\_oid\_str()}
+\begin{verbatim}
+int  ecc_cmp_oid_str(const char *oidstr,
```
please don't make this public API or do you see a future requirement from a user PoV that this is needed?

If you agree, please remove this part and move the function declaration from `tomcrypt_pk.h` to `tomcrypt_private.h`

rmw42 replied:
I put it in the same places as `ecc_get_oid_str()`, but this probably shouldn't be public - it's unlikely that a user application will have created a key but not know what type it is, so this only really makes sense for internal checks.

Will move it and remove the docs.

sjaeckel commented on src/headers/tomcrypt_misc.h:
```diff
@@ -154,6 +154,25 @@ int padding_pad(unsigned char *data, unsigned long length, unsigned long* padded
 #endif  /* LTC_PADDING */
 
+#ifdef LTC_SSH
+typedef enum ssh_data_type_ {
```
~~This should go to `tomcrypt_private.h`~~

Oh no, it shouldn't.

sjaeckel commented on src/headers/tomcrypt_misc.h:
```diff
@@ -154,6 +154,25 @@ int padding_pad(unsigned char *data, unsigned long length, unsigned long* padded
+int ssh_test(void);
+
+/* VA list handy helpers with tuples of <type, data> */
+int ssh_encode_sequence_multi(unsigned char *out, unsigned long *outlen, ...);
```
instead this should probably be as follows?

`int ssh_encode_sequence_multi(unsigned char *out, unsigned long *outlen, int type, void* data, ...);`

not sure though if it would clutter the implementation too much!?

(same applies to `ssh_decode_sequence_multi()`)

rmw42 replied:
Not sure how well va_args lists handle restarting in the middle of the fixed parameters?

I'm going to alter the implementation so we don't have `void *data` being cast all over the place but explicitly collect the correct type from the stack.

sjaeckel replied:
how about something like:

```
diff --git a/src/misc/ssh/ssh_decode_sequence_multi.c b/src/misc/ssh/ssh_decode_sequence_multi.c
index 2507609..3ace026 100644
--- a/src/misc/ssh/ssh_decode_sequence_multi.c
+++ b/src/misc/ssh/ssh_decode_sequence_multi.c
@@ -25,7 +25,7 @@
 */
-int ssh_decode_sequence_multi(const unsigned char *in, unsigned long inlen, ...)
+int ssh_decode_sequence_multi(const unsigned char *in, unsigned long inlen, int type, void *data, ...)
 {
    int           err;
-   ssh_data_type type;
-   void          *data;
+   ssh_data_type t;
+   void          *d;
    va_list       args;
@@ -36,7 +36,6 @@ int ssh_decode_sequence_multi(const unsigned char *in, unsigned long inlen, ...)
    /* Decode values from buffer */
-   va_start(args, inlen);
+   va_start(args, data);
+   t = type;
+   d = data;
    for (;;) {
-      type = (ssh_data_type)va_arg(args, int);
-      data = va_arg(args, void*);
-      LTC_UNUSED_PARAM(data);
 
@@ -46,3 +45,3 @@ int ssh_decode_sequence_multi(const unsigned char *in, unsigned long inlen, ...)
 
-      switch (type) {
+      switch (t) {
          case LTC_SSHDATA_BYTE:
@@ -122,9 +121,12 @@ int ssh_decode_sequence_multi(const unsigned char *in, unsigned long inlen, ...)
          case LTC_SSHDATA_EOL: /* Should never get here */
-            va_end(args);
-            return CRYPT_INVALID_ARG;
+            err = CRYPT_INVALID_ARG;
+            goto out;
        }
+      t = (ssh_data_type)va_arg(args, int);
+      d = va_arg(args, void*);
    }
+out:
    va_end(args);
 
-   return CRYPT_OK;
+   return err;
 }
```

[edit]: FYI yes, that's an incomplete patch ;)

rmw42 replied:
I've gone for something a bit more radical... just updating the tests to cover byte/boolean/uint64 now. Let me know what you think, I can always change it if you're not happy.

rmw42 replied:
Changed to a single return path and re-pushed.

rmw42 replied:
Ah... `warning: second argument to 'va_start' is not the last named parameter [-Wvarargs]`. Since it has to go through the loop twice, `type` can't be in the formal parameters for `ssh_encode_sequence_multi`...

sjaeckel replied:
I guess you forgot to update the `va_start()` argument...

```
diff --git a/src/misc/ssh/ssh_decode_sequence_multi.c b/src/misc/ssh/ssh_decode_sequence_multi.c
index ca088f9..e44162a 100644
--- a/src/misc/ssh/ssh_decode_sequence_multi.c
+++ b/src/misc/ssh/ssh_decode_sequence_multi.c
@@ -25,6 +25,6 @@
 */
-int ssh_decode_sequence_multi(const unsigned char *in, unsigned long inlen, ...)
+int ssh_decode_sequence_multi(const unsigned char *in, unsigned long inlen, int type_, ...)
 {
    int           err;
-   ssh_data_type type;
+   ssh_data_type type = type_;
    va_list       args;
@@ -34,5 +34,4 @@ int ssh_decode_sequence_multi(const unsigned char *in, unsigned long inlen, ...)
    /* Decode values from buffer */
-   va_start(args, inlen);
+   va_start(args, type_);
    for (;;) {
-      type = (ssh_data_type)va_arg(args, int);
 
@@ -103,2 +102,3 @@ int ssh_decode_sequence_multi(const unsigned char *in, unsigned long inlen, ...)
       }
+      type = (ssh_data_type)va_arg(args, int);
    }
```
this compiles w/o warning and tests are OK
(please don't call it `type_` ;) )

rmw42 replied:
That's different to how `der_encode_sequence_multi` etc handle varargs, but I can do it that way if you want?

sjaeckel replied:
OK, leave the signatures as they are, sorry for the noise :)

rmw42 replied:
No worries... it's annoying that there's no elegant solution, it's like an itch in the brain! :)

sjaeckel commented on src/misc/ssh/ssh_decode_sequence_multi.c:
```diff
@@ -0,0 +1,136 @@
+                  return CRYPT_BUFFER_OVERFLOW;
+               }
+               strncpy((char *)data, (const char *)in, size);
+               ((char *)data)[size] = 0; /* strncpy doesn't null terminate */
```
string terminator is `'\0'`

rmw42 replied:
OK

sjaeckel commented on src/misc/ssh/ssh_decode_sequence_multi.c:
```diff
@@ -0,0 +1,136 @@
+               strncpy((char *)data, (const char *)in, size);
+               ((char *)data)[size] = 0; /* strncpy doesn't null terminate */
+            } else {
+               *(char *)data = 0;
```
same

sjaeckel commented on src/misc/ssh/ssh_decode_sequence_multi.c:
```diff
@@ -0,0 +1,136 @@
+            *(unsigned long *)data = size;
+            break;
+         case LTC_SSHDATA_UINT64:
+            size = ((unsigned long)in[0] << 56) |
```
I doubt this works on all architectures [\*cough\*](https://github.com/libtom/libtomcrypt/issues/306#issuecomment-333765095) ;)

sjaeckel replied:
Btw. we also have types for exactly that purpose, `ulong32` and `ulong64` (see `tomcrypt_cfg.h` for details)

rmw42 replied:
Ah, those were exactly what I was looking for. Cheers! Fixing now :)

sjaeckel commented on src/misc/ssh/ssh_encode_sequence_multi.c:
```diff
@@ -0,0 +1,174 @@
+            *out++ = (unsigned char)((size >> 8) & 255);
+            *out++ = (unsigned char)(size & 255);
+            break;
+         case LTC_SSHDATA_UINT64:
```
c.f. comment regarding `LTC_SSHDATA_UINT64` in `ssh_decode_sequence_multi()`

sjaeckel commented on src/misc/ssh/ssh_decode_sequence_multi.c:
```diff
@@ -0,0 +1,136 @@
+                  va_end(args);
+                  return CRYPT_BUFFER_OVERFLOW;
+               }
+               strncpy((char *)data, (const char *)in, size);
```
First I thought why we can't use `strlcpy()`... then I remembered this standardization thingy... as we stay with this please introduce a macro `XSTRNCPY` in `tomcrypt_custom.h` (and also add it to the list in `crypt.tex` and `helper.pl` please)

rmw42 replied:
Done. I've also added a check for `strcpy` in `helper.pl` - not used anywhere in the code as far as I can tell, but presumably it should be alerted if someone introduces it.

sjaeckel commented on src/misc/ssh/ssh_decode_sequence_multi.c:
```diff
@@ -0,0 +1,136 @@
+            *(unsigned long *)data = (size)?1:0;
+            break;
+         case LTC_SSHDATA_UINT32:
+            size = ((unsigned long)in[0] << 24) |
```
can't we use `LOAD32H()` here?

The same applies to all the other cases where we load some `unsigned long` (use `LOAD64H` where appropriate) resp. we could use `STORE{32,64}H()` when storing those values.

rmw42 replied:
Awesome. That's much tidier now :)

sjaeckel commented on tests/ecc_test.c:
```diff
@@ -37,6 +37,93 @@ static unsigned int sizes[] = {
 #endif
 };
 
+const char* names[] = {
```
`static` and a more meaningful name please

rmw42 replied:
That's what it was called when it was function-local ;-)
I'll rename it to `curvenames` and make it static.

sjaeckel commented on tests/ssh_test.c:
```diff
@@ -0,0 +1,237 @@
+   DO(ssh_encode_sequence_multi(buffer, &buflen,
+                                LTC_SSHDATA_UINT32, 0x29b7f4aa,
+                                LTC_SSHDATA_EOL,    NULL));
+   if (XMEMCMP(buffer, uint32, buflen) != 0) return CRYPT_FAIL_TESTVECTOR;
```
You can also use `DO(do_compare_testvector(...`, this prints some more information on why it failed.

... we could probably also add a macro ...

```
#define COMPARE_TESTVECTOR(i, il, s, sl, wa, wi) do { DO(do_compare_testvector((i), (il), (s), (sl), (wa), (wi))); } while(0)
```

rmw42 replied:
Done

sjaeckel commented on src/misc/ssh/ssh_decode_sequence_multi.c:
```diff
@@ -0,0 +1,136 @@
+         case LTC_SSHDATA_BYTE:
+            size = ((unsigned long)in[0]);
+            in++;
+            *(unsigned long *)data = size;
```
why is a byte stored as `unsigned long`?

rmw42 replied:
Managed to outthink myself considering register passing conventions/what this would be on the 'stack'/that it's a cast of `void *` when encoding... basically being silly. I'll make it something sane - uint8?

rmw42 replied:
`unsigned char` then

sjaeckel commented on src/misc/ssh/ssh_decode_sequence_multi.c:
```diff
@@ -0,0 +1,136 @@
+                  return CRYPT_ERROR;
+               }
+            } else {
+               if ((err = mp_read_unsigned_bin(data, (unsigned char *)in, size)) != CRYPT_OK) {
```
~~I just realized that the `src` argument in `unsigned_read()` in `tomcrypt_math.h` could be changed to `const unsigned char *src` as both ltm and tfm nowadays have `const*` arguments where applicable~~ oops ... ltm isn't released yet with those changes... \*cough\* https://github.com/libtom/libtommath/issues/122 \*cough\*

sjaeckel reviewed (changes requested):
Looks good!

:+1:

rmw42 commented:
> in case of `CRYPT_BUFFER_OVERFLOW` we should return required buffer length

Done. I've refactored so it always returns the buffer length (i.e. string length + 1).

sjaeckel commented on helper.pl:
```diff
@@ -53,6 +53,8 @@ sub check_source {
       push @{$troubles->{unwanted_memmove}}, $lineno if $file =~ /^src\/.*\.c$/ && $l =~ /\bmemmove\s*\(/;
       push @{$troubles->{unwanted_memcmp}},  $lineno if $file =~ /^src\/.*\.c$/ && $l =~ /\bmemcmp\s*\(/;
       push @{$troubles->{unwanted_strcmp}},  $lineno if $file =~ /^src\/.*\.c$/ && $l =~ /\bstrcmp\s*\(/;
+      push @{$troubles->{unwanted_strcpy}},  $lineno if $file =~ /^src\/.*\.c$/ && $l =~ /\bstrcpy\s*\(/;
```
:+1:

sjaeckel commented on src/misc/ssh/ssh_decode_sequence_multi.c:
```diff
@@ -0,0 +1,116 @@
+   for (;;) {
+      type = (ssh_data_type)va_arg(args, int);
+
+      if (type == LTC_SSHDATA_EOL) {
```
TBH I prefer a switch-case over a long else-if chain for two reasons: 
1. it's cleaner to read IMO
2. if you do the switch on an enum and don't add a default case, you get immediate feedback from the compiler that the switch has unhandled cases of the enum

rmw42 replied:
I was mostly concerned about the need for a bunch of variable definitions at the top of the function which would each only be used by a single case. What about switch-case with braced code blocks? e.g.:
```
case LTC_SSHDATA_BYTE:
{
   unsigned char *data = va_arg(args, unsigned char *);
...
}
break;
```

sjaeckel replied:
> I was mostly concerned about the need for a bunch of variable definitions at the top of the function which would each only be used by a single case. What about switch-case with braced code blocks?

I don't have a problem with both, but it's true - the braced code blocks worsen readability again

rmw42 replied:
I guess there's not going to be any way to make it pretty :(

Switch-case with variable definitions at the top of the loop, then? And I can put `type` into the formal arg list, change the `for` to `while (type != LTC_SSHDATA_EOL)` and put the reading of `type` at the end of the block. Sound good?

Any other changes I should make after that?

sjaeckel replied:
> I guess there's not going to be any way to make it pretty :(

unfortunately not :/



> Switch-case with variable definitions at the top of the loop, then?

I'd say at the top of the function, not the loop



> And I can put `type` into the formal arg list, change the `for` to `while (type != LTC_SSHDATA_EOL)` and put the reading of `type` at the end of the block. Sound good?

why not change the signature to the proposed `int ssh_decode_sequence_multi(const unsigned char *in, unsigned long inlen, int type, void *data, ...)`? 

Should it really be a valid case to call `ssh_decode_sequence_multi(foo, bar, LTC_SSHDATA_EOL)`?

rmw42 replied:
I can't put `type` into the formal parameters because the encoding needs to go through the list twice, and can't be `va_start`ed from the middle of the formal parameters, so it would make the signatures of the two methods different :(
And `data` shouldn't really go in there because it then forces things to be `void*` pointers even when they should be integers :(

sjaeckel replied:
Both valid points I didn't see!

Again, sorry for the noise!

sjaeckel replied:
But please, still change back to switch-case :)

rmw42 replied:
Done, and I've moved the params to the top of the function.

rmw42 commented:
I added a new check in `ssh_decode_sequence_multi` to ensure we don't read past the end of the buffer, which caught a bug in the name-list decoding tests (I'd used `sizeof(string)` instead of `sizeof(nlist3)` - copy/paste bug).

rmw42 commented:
OK, I've rebased to current develop branch (might need to do so again with the `ecc_verify_hash_ex` reversion), and I think that's all the things we've discussed... if there are any more issues, please let me know 😄

sjaeckel commented on src/headers/tomcrypt_misc.h:
```diff
@@ -154,6 +154,25 @@ int padding_pad(unsigned char *data, unsigned long length, unsigned long* padded
+   LTC_SSHDATA_EOL
+} ssh_data_type;
+
+int ssh_test(void);
```
oops, just found that this doesn't belong here anymore but in `tests/tomcrypt_test.h`

rmw42 replied:
Moved.

sjaeckel commented on src/headers/tomcrypt_pk.h:
```diff
@@ -251,7 +251,9 @@ typedef enum ecc_signature_type_ {
-   LTC_ECCSIG_ETH27      = 0x2
+   LTC_ECCSIG_ETH27      = 0x2,
+   /* SSH + ECDSA signature format defined by RFC5656 */
+   LTC_ECCSIG_RFC5656    = 0x3
```
I'm not 100% sure if this shouldn't be within `#ifdef LTC_SSH`?

if we're already at it, can you please directly add a `,` at the end :)

rmw42 replied:
I think it makes sense to _define_ the signature format in all cases, whether or not the implementation is built into the library. If `LTC_SSH` isn't defined, then it'll fall through to the "unknown format" errors in the functions.

I think a trailing comma in an enum results in a compiler warning. Not sure, though?

sjaeckel replied:
> I think it makes sense to _define_ the signature format in all cases

agreed!



> I think a trailing comma in an enum results in a compiler warning. Not sure, though?

shouldn't on modern compilers :) (e.g. the `ltc_asn1_XX` enums have it as well)

rmw42 replied:
> shouldn't on modern compilers

Yup. I thought I'd seen this somewhere recently, but apparently not - it compiles cleanly.

sjaeckel commented on src/misc/ssh/ssh_decode_sequence_multi.c:
```diff
@@ -0,0 +1,120 @@
+      }
+
+      /* Check we've not read too far */
+      if (in > sentinel) {
```
:heart:

sjaeckel replied:
shouldn't this kind of check be in each case where we read more than 1 byte?

rmw42 replied:
Probably, but it would make the implementation a lot uglier. It'd mean we couldn't use the `LOAD32H` macros as they don't do checks after each byte.

The main reason I could see for doing that is that there might be information leaked if it can read past the end of the buffer... but only by someone with the ability to change the code, who could read from that memory anyway.

It's in the loop, so it will be checked after each object read, whether that's 4 or 8 or however many bytes. Personally, I think that's OK, but I'm happy to move it if you disagree.

sjaeckel replied:
> it would make the implementation a lot uglier

sure, but it's required IMO as the following isn't the only bad thing that could happen IMO

> The main reason I could see for doing that is that there might be information leaked if it can read past the end of the buffer...

... but if you put the buffer at the end of a page it could also segfault or not?

And isn't it even UB to read past an allocated buffer?

rmw42 replied:
Well, the same would apply to _anything_ that reads from a buffer if the specified length is greater than the size of the buffer... it shouldn't segfault unless someone has done something weird with the virtual memory page protection. Even if it does, if an application asks us to read 1,000,000 bytes from a 1,000 byte buffer, or passes us a pointer which cannot be read from at all, is a segfault not an appropriate response?

Instead of a sentinel at the end, I could keep track of the bytes available to read, and put a second switch-case in the loop to calculate the size which will be read? That's still going to be somewhat delicate around the variable-length string/mpint cases...

rmw42 replied:
OK, I've ended up with three switch-cases:
- the length field of the object (if needed)
- read/calculate the length of the object
- read the object itself

Should now be impossible for `ssh_decode_sequence_multi` to read beyond the end of the input.

sjaeckel replied:
> is a segfault not an appropriate response?

sure it is and if we can prevent it by properly checking indicators like the required vs available length we should do it IMO.



> Instead of a sentinel at the end, I could keep track of the bytes available to read, and put a second switch-case in the loop to calculate the size which will be read? That's still going to be somewhat delicate around the variable-length string/mpint cases...

how about

```
diff --git a/src/misc/ssh/ssh_decode_sequence_multi.c b/src/misc/ssh/ssh_decode_sequence_multi.c
index 2375df8..aa171b3 100644
--- a/src/misc/ssh/ssh_decode_sequence_multi.c
+++ b/src/misc/ssh/ssh_decode_sequence_multi.c
@@ -37,14 +37,21 @@ int ssh_decode_sequence_multi(const unsigned char *in, unsigned long inlen, ...)
    unsigned long size, bufsize;
+   long len;
 
    LTC_ARGCHK(in    != NULL);
+   LTC_ARGCHK(inlen <= LONG_MAX);
 
    /* Decode values from buffer */
+   len = inlen;
    va_start(args, inlen);
    while ((type = (ssh_data_type)va_arg(args, int)) != LTC_SSHDATA_EOL) {
+      err = CRYPT_BUFFER_OVERFLOW;
       switch (type) {
          case LTC_SSHDATA_BYTE:
+            if (len < 1) goto error;
             cdata = va_arg(args, unsigned char*);
             *cdata = *in++;
+            len--;
             break;
          case LTC_SSHDATA_BOOLEAN:
+            if (len < 1) goto error;
             cdata = va_arg(args, unsigned char*);
@@ -55,4 +62,6 @@ int ssh_decode_sequence_multi(const unsigned char *in, unsigned long inlen, ...)
             *cdata = (*in++)?1:0;
+            len--;
             break;
          case LTC_SSHDATA_UINT32:
+            if (len < 4) goto error;
             u32data = va_arg(args, ulong32*);
@@ -60,4 +69,6 @@ int ssh_decode_sequence_multi(const unsigned char *in, unsigned long inlen, ...)
             in += 4;
+            len -= 4;
             break;
          case LTC_SSHDATA_UINT64:
+            if (len < 8) goto error;
             u64data = va_arg(args, ulong64*);
@@ -65,2 +76,3 @@ int ssh_decode_sequence_multi(const unsigned char *in, unsigned long inlen, ...)
             in += 8;
+            len -= 8;
             break;
@@ -73,6 +85,3 @@ int ssh_decode_sequence_multi(const unsigned char *in, unsigned long inlen, ...)
             if (size > 0) {
-               if (size >= bufsize) {
-                  err = CRYPT_BUFFER_OVERFLOW;
-                  goto error;
-               }
+               if (size >= bufsize) goto error;
                XSTRNCPY(sdata, (const char *)in, size);
@@ -83,4 +92,6 @@ int ssh_decode_sequence_multi(const unsigned char *in, unsigned long inlen, ...)
             in += size;
+            len -= size;
             break;
          case LTC_SSHDATA_MPINT:
+            if (len < 4) goto error;
             vdata = va_arg(args, void*);
@@ -91,2 +102,3 @@ int ssh_decode_sequence_multi(const unsigned char *in, unsigned long inlen, ...)
             } else {
+               if (len < size) goto error;
                if ((err = mp_read_unsigned_bin(vdata, (unsigned char *)in, size)) != CRYPT_OK)          { goto error; }
@@ -94,2 +106,3 @@ int ssh_decode_sequence_multi(const unsigned char *in, unsigned long inlen, ...)
             in += size;
+            len -= size;
             break;
```

rmw42 replied:
Yeah, that'd work too, but some of the variable-length ones would need extra checks.

How's the triple-switch version? It's a bit longer but maybe clearer intent...

sjaeckel replied:
> How's the triple-switch version?

I'm not too much a fan of it, @karel-m your thoughts?

sjaeckel commented on src/misc/ssh/ssh_encode_sequence_multi.c:
```diff
@@ -0,0 +1,167 @@
+
+   /* Check we have sufficient space */
+   if (*outlen < size) {
+      err = CRYPT_BUFFER_OVERFLOW;
```
didn't we say we should return the required size in this error-case?

rmw42 replied:
Oops. Changed another case (`ecc_ssh_decode_name`) but missed this one. Fixing it now.

sjaeckel commented on src/pk/ecc/ecc_cmp_oid_str.c:
```diff
@@ -0,0 +1,40 @@
+
+   LTC_ARGCHK(key != NULL);
+
+   if ((err = pk_oid_num_to_str(key->dp.oid, key->dp.oidlen, oid, &oidlen)) != CRYPT_OK)                { goto error; }
```
if we merge this after #404 there's going to be `int pk_oid_cmp_with_ulong(const char *o1, const unsigned long *o2, unsigned long o2size);` ;)

rmw42 replied:
I can drop `ecc_cmp_oid_str` entirely, then. Turned out that it was only needed for the secp256k1 checks for `LTC_ECCSIG_ETH27`. Let me know if we want to do that.

sjaeckel replied:
We keep it for now as is I'd say as the other PR seems to take some more iterations as it's not entirely ready yet.

sjaeckel replied:
Even better, simply cherry-pick those commits and whoever gets merged first will have them and the second will be rebased and the commits removed :)  Git FTW

c0407783ac383f9509a62dde99de4b6606c4d4f7
c81e1cb1dc8b36286233605c3a10c47584ec0e79
6abfab4d03389ccfe2c31f7ead84c706befc3e60

rmw42 replied:
Took a while to figure out the syntax - thought I'd have to name the branch I was pulling from - but that was remarkably painless! 🍒 

I'll get rid of `ecc_cmp_oid_str` now.

sjaeckel commented on tests/common.h:
```diff
@@ -17,10 +17,12 @@ extern prng_state yarrow_prng;
 #define DO(x) do { run_cmd((x), __LINE__, __FILE__, #x, NULL); } while (0)
 #define DOX(x, str) do { run_cmd((x), __LINE__, __FILE__, #x, (str)); } while (0)
 #define SHOULD_FAIL(x) do { run_cmd((x) != CRYPT_OK ? CRYPT_OK : CRYPT_FAIL_TESTVECTOR, __LINE__, __FILE__, #x, NULL); } while (0)
+#define COMPARE_TESTVECTOR(i, il, s, sl, wa, wi) do { DO(do_compare_testvector((i), (il), (s), (sl), (wa), (wi))); } while(0)
```
that define is only required once after the `#endif`

rmw42 replied:
OK

sjaeckel reviewed (commented):
Sorry for more nitpicks, minor comments, but there's also one important thing I only spotted now (regarding the length-checks when decoding)

rmw42 committed:
Update makefiles

rmw42 commented:
I've updated the length-check to remove one of the switches and keep track of the length remaining, rather than the end pointer, and rebased to the `develop` branch to fix the merge issues around cherry-picking `pk_oid_cmp_with_ulong`.

sjaeckel commented:
@mkj did you have a look at this?

mkj commented on src/misc/ssh/ssh_decode_sequence_multi.c:
```diff
@@ -0,0 +1,156 @@
+               XSTRNCPY(sdata, (const char *)in, size);
+               sdata[size] = '\0'; /* strncpy doesn't NUL-terminate */
+            } else {
+               *sdata = '\0';
```
Invalid write if size==0 && bufsize==0

rmw42 replied:
Moved the `size >= bufsize` check up, out of the `if (size > 0)` case. That way if the string is 0 bytes long - an empty string - we still need a buffer of at least 1 byte to output a C-style NUL-terminated string.

mkj commented on src/misc/ssh/ssh_decode_sequence_multi.c:
```diff
@@ -0,0 +1,156 @@
+            if (size == 0) {
+               if ((err = mp_set(vdata, 0)) != CRYPT_OK)                                                { goto error; }
+            } else {
+               if ((err = mp_read_unsigned_bin(vdata, (unsigned char *)in, size)) != CRYPT_OK)          { goto error; }
```
Check that the top bit isn't set, that implies a negative value
https://tools.ietf.org/html/rfc4251#page-8

rmw42 replied:
Good catch. Done.

mkj reviewed (commented):
Looks OK, couple of small things I've noted.

Unrelated to the patch, I'm curious where SSH signature encoding is being used externally to the SSH protocol?

rmw42 commented:
> Unrelated to the patch, I'm curious where SSH signature encoding is being used externally to the SSH protocol?

I don't think it is used anywhere else... @karel-m asked for this so you could avoid having to transform another signature format in Dropbear.

rmw42 committed:
Fixes suggested by Matt Johnston

rmw42 committed:
GCC 4.8 -Wmaybe-uninitialized is *really* dumb

karel-m commented:
@mkj the idea was mine. During my (unfinished) work - https://github.com/karel-m/dropbear/commit/4530ff68975932680d674a33ea477fa7afc79ade - on a PR adopting Dropbear to the new libtomcrypt ecc_* stuff I saw the dark magic in dropbear's `ecc.c` related to ECDSA signatures which lead me to the idea that libtomcrypt should do a better job supporting our "prominent customer" (= dropbear).

mkj commented:
Ah right. I'm not sure it really makes sense for libtomcrypt to be parsing SSH packets when it's alongside another parser anyway. 

I think it would work fine if there were just a 
```
int ecc_verify_hash_raw(void *r, void *s,
    const unsigned char *hash, unsigned long hashlen,
    int *stat, const ecc_key *key)
```

etc which could be called? Dropbear needs to do some work outside anyway to know which hash_desc to use.

Currently Dropbear's bundled libtomcrypt skips LTC_DER since it isn't necessary (I just treat them as opaque blobs in a few places) - it avoids some code size. It would be nice to avoid having to pull that in too, maybe the `dropbear_ecc_curve.oid` in the new branch should just be a long array to memcmp()?

sjaeckel commented:
> maybe the `dropbear_ecc_curve.oid` in the new branch should just be a long array to memcmp()?

The plan is to make all OID appearances per default string-based (providing still the possibility to use the ulong-array approach to de-/encode via the DER API). Would that be fine for you as well?

rmw42 commented:
Rebased to current develop branch

karel-m commented:
Ad @mkj 's comments

> I think it would work fine if there were just a `int ecc_verify_hash_raw(void *r, void *s, ...)`

I do not like this. This will expose internal libtomcrypt guts (pointers to bignum structures) which we do not have yet in any public API function. Plus there is of course a completely new bunch of troubles related to the fact that math provider is not always libtommath.

> Currently Dropbear's bundled libtomcrypt skips LTC_DER since it isn't necessary...

This is a valid point. The current all-in-one `ecc_sign_hash_ex` and `ecc_verify_hash_ex` forces anyone using ECDSA to link the whole ASN.1 stuff to their binary even if they are using just RFC 7518 or RFC 5656 (ssh2) signature format. IMO it is worth rethinking in a separate PR.

@sjaeckel can we merge this PR? (to me it looks good)

rmw42 commented:
> The current all-in-one `ecc_sign_hash_ex` and `ecc_verify_hash_ex` forces anyone using ECDSA to link the whole ASN.1 stuff to their binary

We could put that branch of the conditional on `#ifdef LTC_DER` so `LTC_ECCSIG_ANSIX962` is only available as a signature format if the library is built with DER support and returns an error otherwise.

karel-m commented:
This will not handle a static linking scenario when you have a static library built with all ASN.1/DER support but you want to just use RFC 7518 ECDSA signatures. I think we should somehow split `ecc_sign_hash_ex.c` and `ecc_verify_hash_ex.c` into more `.c` files.

rmw42 commented:
Cheers! 👍
